pimd/pim_pim.c

```diff
@@ -585,22 +585,67 @@ static int pim_msg_send_frame(int fd, char *buf, size_t len,
 		return pim_msg_send_frame(fd, (char *)ip2, sendlen, dst, salen,
 					  ifname);
 	}
-#endif
 
 	zlog_warn(
 		"%s: sendto() failure to %pSU: iface=%s fd=%d msg_size=%zd: %m",
 		__func__, dst, ifname, fd, len);
 	return -1;
 }
 
+#else
+static int pim_msg_send_frame(pim_addr src, pim_addr dst, ifindex_t ifindex,
+			      struct iovec *message, int fd)
+{
+	int retval;
+	struct msghdr smsghdr = {};
+	struct cmsghdr *scmsgp;
+	union cmsgbuf {
+		struct cmsghdr hdr;
+		uint8_t buf[CMSG_SPACE(sizeof(struct in6_pktinfo))];
+	};
+	struct in6_pktinfo *pktinfo;
+	struct sockaddr_in6 dst_sin6 = {};
+
+	union cmsgbuf cmsg_buf = {};
+
+	/* destination address */
+	dst_sin6.sin6_family = AF_INET6;
+#ifdef SIN6_LEN
+	dst_sin6.sin6_len = sizeof(struct sockaddr_in6);
+#endif /*SIN6_LEN*/
+	dst_sin6.sin6_addr = dst;
+	dst_sin6.sin6_scope_id = ifindex;
+
+	/* send msg hdr */
+	smsghdr.msg_iov = message;
+	smsghdr.msg_iovlen = 1;
+	smsghdr.msg_name = (caddr_t)&dst_sin6;
+	smsghdr.msg_namelen = sizeof(dst_sin6);
+	smsghdr.msg_control = (caddr_t)&cmsg_buf.buf;
+	smsghdr.msg_controllen = CMSG_SPACE(sizeof(struct in6_pktinfo));
+
+	scmsgp = CMSG_FIRSTHDR(&smsghdr);
+	scmsgp->cmsg_level = IPPROTO_IPV6;
+	scmsgp->cmsg_type = IPV6_PKTINFO;
+	scmsgp->cmsg_len = CMSG_SPACE(sizeof(struct in6_pktinfo));
+
+	pktinfo = (struct in6_pktinfo *)(CMSG_DATA(scmsgp));
+	pktinfo->ipi6_ifindex = ifindex;
+	pktinfo->ipi6_addr = src;
```

Comment:
So for pim register msg the src IP cannot be a local link address otherwise the RP wont be able to reply back. So you have to handle the pim register packet separately. If its a pim register packet set the src address as 0 so that the kernel fills it.

Reply:
Fixed